pkg/listwatch/listwatch.go

```diff
@@ -154,11 +155,13 @@ func (mlw multiListerWatcher) List(options metav1.ListOptions) (runtime.Object,
 		for _, item := range items {
 			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
 		}
-		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+		if !resourceVersions.Has(metaObj.GetResourceVersion()) {
+			resourceVersions.Insert(metaObj.GetResourceVersion())
+		}
 	}
 	// Combine the resource versions so that the composite Watch method can
 	// distribute appropriate versions to each underlying Watch func.
-	l.ListMeta.ResourceVersion = strings.Join(resourceVersions, "/")
+	l.ListMeta.ResourceVersion = strings.Join(resourceVersions.List(), "/")
```

Comment:
I don't understand the intent here.  ResourceVersion for a list is a singular value.  Placing `val/val/val/val` doesn't seem like it would ever have worked correctly.  While I can appreciate that interpreting as a numeric is disallowed by API *users*, it is allowed by the API storage layer since the storage layer knows that the only valid resourceVersion for etcd is a numeric.

Reply:
IIRC, the idea here was to create a synthetic resource version that our custom watch implementation knows how to split and fan out into several different list calls. Note that this synthesized resource version is never supposed to actually hit the API; it is abused by the custom multi-list-watch implementation to combine multiple different lists into one single state.